README.md
# KNP Anthrax Model

This configurable agent-based model simulates movement patterns and behaviors of spatially explicit Kudus and Impalas in a georeferenced environment (by default, the northern half of the Kruger National Park (KNP)) to enable analysis of the distribution and density of Anthrax pathogens. The model runs on the [Multi-Agent Research and Simulation (MARS) Framework](https://mars-group-haw.github.io/index.html).

| Simulation area: Northern KNP                  | Kudu and Impala movement density                                  | Anthrax infection site density                                                      |
| ---------------------------------------------- | ------------------------------------------------------------ | ---------------------------------------------- |
| ![Simulation area](./docs/Simulation_Area.png) | ![Kudu and Impala movement density](./docs/Impala_and_Kudu_Movement.png) | ![Anthrax infection site density](./docs/Anthrax_Density.png)                          |

## Project Setup and Usage

There are two options for setting up and using the project. **Option 1** allows you to see and change the source code as well as configure and run the model. **Option 2** allows you to only configure and run the model without being able to see or change the source code.

### Option 1: Source Code

1. Install the [.NET SDK 6.0](https://dotnet.microsoft.com/en-us/download), either manually or within an interactive development environment (IDE) like [JetBrains Rider](https://www.jetbrains.com/de-de/rider/).
2. Copy the project from GitHub to your machine in one of the following ways:
   - Clone the project by entering the command `git clone https://github.com/MARS-Group-HAW/knp-anthrax-model.git` (**Note:** to do this, you must have [Git](https://git-scm.com/downloads) installed).
   - Download the project by clicking the "Code" button and selecting "Download ZIP".
3. Configure the model as needed (see [Configuration](#configuration) section below).
4. Build and run the project in one of the following ways:
   - In a terminal:
     1. Open a terminal and navigate to the folder `KNPAnthrax/`.
     3. Build the project by entering the command `dotnet build`.
     4. Run the project by enter the command `dotnet run`.
   - In the Rider IDE:
     1. Open Rider and open the solution file `KNPAnthrax.sln`. This opens the project in Rider.
     2. Press the "Run"-Button in the top-right area of the window.

### Option 2: Executable

Use a self-contained and executable [simulation box](https://github.com/MARS-Group-HAW/knp-anthrax-model/releases) to configure and run the model.

1. At the above link, download the box that fits your operating system.
2. Unzip the box.
3. Configure the model as needed (see [Configuration](#configuration) section below).
3. Run the `KNPAnthrax` executable.

## Configuration

To configure the model components (global simulation variables, agents, and layers), change the values in the file `config.json`, which is located in the folder `KNPAnthrax`. The most important configuration options are described below.

### Global simulation parameters

Simulation start and end time:
- `startPoint` specifies the start time
- `endPoint` specifies the end time

Tick lenght: 8 hours
- `deltaT` specifies the number
- `detlaTUnit` specifies the unit

### Agent parameters

Each agent starts with a given `Energy` property. During each tick, the value of `Energy` decreases by `1`.
- While the `Energy` value is greater than or equal to `15`, the agent will randomly walk arround while obeying certain constraints (landtype preference and proximity to water source).
- When the `Energy` value is below `15`, the agent begins to search for and move to the nearest water source. Upon reaching a water source, the agent drinks and stays there until `Energy` is at `100`.

#### `Kudu`

Preferred landscape type: Woodland

| Key name                                                   | Value type and value range                             | Comment                                                      |
| ------------------------------------------------------------ | ---------------------------------------------------- | ------------------------------------------------------------ |
| `count`                                                      | `int`, at least `0`                                  | Number of agents, ⚠️ in the file `Resources/impala.csv`, you need the same number of rows. |
| `SpawnMinEnergy`, `SpawnMaxEnergy`                           | `int`, between `0` and `100`                         | Defines a range in which the `Energy` of each agent will be at spawning. This allows for more dynamic animal behaviour, so that not all animals need water at the exact same time. |
| `SpawnWoodlandProbability`, `SpawnSavannaProbability`        | `float`, each between `0.0` and `1.0`, sum needs to be `1.0` | Probability of thes agents spawning in Woodland or Savanna land type.       |
| `MinMovementPerTickInM`, `MaxMovementPerTickInM`             | `int`, in meters                                     | Distance in meters that the agent moves per tick will be between these values. |
| `MovementOnWoodland`                                         | `float`, between `0.0` to `1.0`, sum needs to be `1.0`                                               | Currently has no effect, since the Kudu prefers Woodland                |
| `MovementOnSavanna`                                          | `float`, between `0.0` to `1.0`, sum needs to be `1.0`                                               | If the agent reaches a Savanna patch, it will continue to walk on it with this probability. |
| `MaxDistanceFromWaterInM`                                    | `int` in meters                                      | Maximum distance the agent will move away from the nearest water source. |
| `AnthraxInfectionProbability`                                | `float`, between `0.0` and `1.0`                                              | This probability is multiplied with values from the `AnthraxLayer` grid cells to determine whether the agent becomes infected. |
| `MinInfectionDurationInTicks`, `MaxInfectionDurationInTicks` | `int`, in number of ticks                            | Range of number of ticks that an infected agent will leave a trace on the `AnthraxLayer` (a trace represents a new Anthrax infection site caused by a dead animal). |
| `OutputAgentTrack`                                           | `bool`                                               | If set to `true`, each agent produces a single GeoJSON file containing its movement trajectory throughout the simulation. |

#### `Impala`

Preferred landscape types: Woodland, Savanna

| Key name                                                   | Value type and value range                                                | Comment                                                      |
| ------------------------------------------------------------ | ---------------------------------------------------- | ------------------------------------------------------------ |
| `count`                                                      | `int`, at least `0`                                  | Number of agents, ⚠️ in the file `Resources/impala.csv`, you need the same number of rows. |
| `SpawnMinEnergy`, `SpawnMaxEnergy`                           | `int`, between `0` and `100`                         | Defines a range in which the `Energy` of each agent will be at spawning. This allows for more dynamic animal behaviour, so that not all animals need water at the exact same time. |
| `SpawnWoodlandProbability`, `SpawnSavannaProbability`        | `float`, each between `0.0` and `1.0`, sum needs to be `1.0` | Probability of thes agents spawning in Woodland or Savanna land type.       |
| `MinMovementPerTickInM`, `MaxMovementPerTickInM`             | `int`, in meters                                     | Distance in meters that the agent moves per tick will be between these values. |
| ⚠️ no Woodland/Savanna                                       | --                                                   | No options are given due to the near-equal disitribution (60%-40%).    |
| `MaxDistanceFromWaterInM`                                    | `int` in meters                                      | Maximum distance the agent will move away from the nearest water source. |
| `AnthraxInfectionProbability`                                | `float`, between `0.0` and `1.0`                     | This probability is multiplied with values from the `AnthraxLayer` grid cells to determine whether the agent becomes infected. |
| `MinInfectionDurationInTicks`, `MaxInfectionDurationInTicks` | `int`, in number of ticks                            | Range of number of ticks that an infected agent will leave a trace on the `AnthraxLayer` (a trace represents a new Anthrax infection site caused by a dead animal). |
| `OutputAgentTrack`                                           | `bool`                                               | If set to `true`, each agent produces a single GeoJSON file containing its movement trajectory throughout the simulation. |

### Layer parameters

#### `LandscapeLayer`

Provides landtype classification. Input data are manually mapped to `Woodland` and `Savanna` categories.

#### `Perimeter`

Simulation area, derived from union of landscape layer input.

#### `WaterLayer`

Waterholes and rivers, derived from various water data sources. The GeoJSON file can contain POINTs and LINESTRINGs.

> **Warning**
>
> All geometries given in the `WaterLayer` **must** be inside the `Perimeter` shape, or the simulation will break.

#### `AnthraxLayer`

Raster data with 1x1km resolution containing Anthrax carcass sites from the northern half of the KNP.

## Outputs

Depending on how you are using the project (see the section [Project Setup and Usage](#project-setup-and-usage) above), there are two ways to analyze and visualize results.

### Option 1

When running the project via a terminal or from within the Rider IDE (see the section [Option 1](#option-1-source-code) above), output files will be placed in the folder `KNPAnthrax/bin/Debug/net6.0/`. These output files can be visualized with [kepler.gl](https://kepler.gl) via drag-and-drop.

- `Kudu_trips.geojson` and `Impala_trips.geojson` contain time series movement data of each agent type.
  - **Note:** Due to the 8h time resulation and the corresponding large movement distances per tick, kepler.gl is unable to visualize the movement well.
  - **Note:** The file is only generated when `{"kind": "trips"}` is set, be aware that for long simulation/large agen cpunt this file might get to large to be rendered by the system.
- `Impala.csv` and `Kudu.csv` will contain all agent attributes, notable are the following columns:
  - `Tick`: Tck of simulation
  - `DateTime`: current time of the simulation
  - `InfectedCounter`: currently infected inside the agent/herd
  - `InfectedTotalCounter`: total amount of infections inside the agent/herd
  - See `Analysis/CSVAnalsis.ipnyb` for example visualisations.
  - **Note**: only ticks/rows will be created for the first/last tick and if something for the infection related columns will change (to reduce file size/simulation time)
- `LandscapeLayer_types.geojson` contains the mapping of the input land type file to the repsective landtype in the model (Woodland, Savanna).
- `AnthraxLayer_Start.geojson` is the input raster data of the Anthrax distriubtion (the input ASC/GeoTIFF converted to GeoJSON).
- `AnthraxLayer_End.geojson` is a heatmap showing the anthrax densitiy after all animals have left their traces.
- `KuduMovement.geojson` and `ImpalaMovement.geojson` are heatmaps showing the movement densities of all agents. Each agent increments the value of the cell in which it stands by `0.1`.

**Note:** [kepler.gl](https://kepler.gl/) does not enable saving configurations or visualizations. For convenience, you can also use the provided Jupyter Notebook (see the folder `Analysis/`) with a preset configuration to load some data. To do so, start a [Docker Container](https://www.docker.com/products/docker-desktop/) within which the Jupyter Hub will run. To start the Docker container, follow these steps:
1. Install and start Docker Desktop.
2. Open a terminal and navigate to the folder `Analysis/`.
3. Start the Docker container by entering one of the following commands:
   - For Mac: `./notebookdocker.sh`
   - For Windows: `./notebookdocker.bat`
3. In your internet browser, enter `localhost:8888` in your address bar.
4. Open and run the Jupyter notebook `Analysis.ipynb`. The kepler.gl configurations will be stored in the file `map_config.py` for later use.

### Option 2

When using one of the executable boxes (see the section [Option 2](#option-2-executable) above), the output files will be located in the same folder as the `KNPAnthrax` executable file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.2](https://github.com/MARS-Group-HAW/knp-anthrax-model/compare/v0.1.1...v0.1.2) (2023-02-24)


### Bug Fixes

* check for preferred land type before random walk reduces lookups, as well as fixes infinite loop for Kudu stuck on unkown land type ([ef1e165](https://github.com/MARS-Group-HAW/knp-anthrax-model/commit/ef1e1654e21a7cc0861838e4f0e108b816ee6bfe))
* only keep list of positions if agent linetrack output is enabled for performance ([2de466f](https://github.com/MARS-Group-HAW/knp-anthrax-model/commit/2de466f96d5138f769fbefdcbec8a825dcf5f7df))


KNPAnthrax/Model/Kudu.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Mars.Common;
using Mars.Common.Core.Random;
using Mars.Components.Layers;
using Mars.Interfaces.Agents;
using Mars.Interfaces.Annotations;
using Mars.Interfaces.Environments;
using Mars.Interfaces.Layers;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using NetTopologySuite.Features;
using NetTopologySuite.Geometries;
using NetTopologySuite.IO;
using NetTopologySuite.Operation.Distance;
using Position = Mars.Interfaces.Environments.Position;

namespace KNPAnthrax.Model;


public class Kudu : IAgent<AnimalLayer>, IPositionable
{
    #region Properties and Fields

    /// <summary>
    ///     Enables the agent's state at the end of the current tick to be persisted to a CSV output file.
    /// </summary>
    public bool StoreTickResult { get; set; }
    
    /// <summary>
    ///     The latitude of the current geo-referenced position of the agent.
    /// </summary>
    [PropertyDescription(Name = "Latitude")]
    public double Latitude { get; set; }

    /// <summary>
    ///     The longitude of the current geo-referenced position of the agent.
    /// </summary>
    [PropertyDescription(Name = "Longitude")]
    public double Longitude { get; set; }
    
    /// <summary>
    ///     The current energy level of the agent.
    /// </summary>
    [PropertyDescription(Name = "Energy")]
    public double Energy { get; set; }
    
    /// <summary>
    ///     The minimum energy level of the agent upon being spawned.
    /// </summary>
    [PropertyDescription(Name = "SpawnMinEnergy")]
    public double SpawnMinEnergy { get; set; }
    
    /// <summary>
    ///     The maximum energy level of the agent upon being spawned.
    /// </summary>
    [PropertyDescription(Name = "SpawnMaxEnergy")]
    public double SpawnMaxEnergy { get; set; }
    
    /// <summary>
    ///     The agent's current state.
    /// </summary>
    [PropertyDescription(Name = "State")]
    public AnimalState State { get; set; }

    /// <summary>
    ///     The probability of being spawned on "woodland" land type.
    /// </summary>
    [PropertyDescription(Name = "SpawnWoodlandProbability")]
    public double SpawnWoodlandProbability { get; set; }

    /// <summary>
    ///     The probability of being spawned on "savanna" land type.
    /// </summary>
    [PropertyDescription(Name = "SpawnSavannaProbability")]
    public double SpawnSavannaProbability { get; set; }
    
    /// <summary>
    ///     The minimum movement distance in meters per tick.
    /// </summary>
    [PropertyDescription(Name = "MinMovementPerTickInM")]
    public double MinMovementPerTickInM { get; set; }

    /// <summary>
    ///     The maximum movement distance in meters per tick.
    /// </summary>
    [PropertyDescription(Name = "MaxMovementPerTickInM")]
    public double MaxMovementPerTickInM { get; set; }
    
    /// <summary>
    ///     The relative preference towards moving on "woodland" land type.
    /// </summary>
    [PropertyDescription(Name = "MovementOnWoodland")]
    public double MovementOnWoodland { get; set; }

    /// <summary>
    ///     The relative preference towards moving on "savanna" land type.
    /// </summary>
    [PropertyDescription(Name = "MovementOnSavanna")]
    public double MovementOnSavanna { get; set; }
    
    /// <summary>
    ///     The maximum distance in meters allowed between the agent and the nearest water source. 
    /// </summary>
    [PropertyDescription(Name = "MaxDistanceFromWaterInM")]
    public double MaxDistanceFromWaterInM { get; set; }
    
    /// <summary>
    ///     The probability of becoming infected with Anthrax when in a cell that contains Anthrax.
    /// </summary>
    [PropertyDescription(Name = "AnthraxInfectionProbability")]
    public double AnthraxInfectionProbability { get; set; }

    /// <summary>
    ///     The minimum number of ticks that an Anthrax infection lasts (from exposure to death).
    /// </summary>
    [PropertyDescription(Name = "MinInfectionDurationInTicks")]
    public int MinInfectionDurationInTicks { get; set; }

    /// <summary>
    ///     The maximum number of ticks that an Anthrax infection lasts (from exposure to death).
    /// </summary>
    [PropertyDescription(Name = "MaxInfectionDurationInTicks")]
    public int MaxInfectionDurationInTicks { get; set; }
    
    /// <summary>
    ///     Enables the agent's movement trajectory to be written to a GeoJSON file.
    /// </summary>
    [PropertyDescription(Name = "OutputAgentTrack")]
    public bool OutputAgentTrack { get; set; }
    
    /// <summary>
    ///     The layer on which the agent lives.
    /// </summary>
    private AnimalLayer Layer { get; set; }
    
    /// <summary>
    ///     The perimeter of the simulation environment.
    /// </summary>
    [PropertyDescription(Name = "LandscapeLayer")]
    public LandscapeLayer LandscapeLayer { get; set; }
    
    /// <summary>
    ///     A layer that holds water sources in the form of vector features.
    /// </summary>
    [PropertyDescription(Name = "WaterLayer")]
    public WaterLayer WaterLayer { get; set; }
    
    /// <summary>
    ///     A grid-based layer that holds Anthrax infection sites.
    /// </summary>
    [PropertyDescription(Name = "AnthraxLayer")]
    public AnthraxLayer AnthraxLayer { get; set; }
    
    /// <summary>
    ///     The perimeter of the simulation environment.
    /// </summary>
    [PropertyDescription(Name = "Perimeter")]
    public Perimeter Perimeter { get; set; }

    /// <summary>
    ///     A grid-based layer that is used to store movement locations of Kudu agents.
    /// </summary>
    [PropertyDescription(Name = "KuduMovement")]
    public KuduMovement KuduMovement { get; set; }
    
    /// <summary>
    ///     A collection of land types that the agent prefer to be on.
    /// </summary>
    private readonly List<LandscapeType> _preferredLandTypes = new() { LandscapeType.Woodland };
    
    /// <summary>
    ///     A counter to keep track of the number of infections this agent has had.
    /// </summary>
    public int InfectedCounter { get; set; }
    
    /// <summary>
    ///     A collection that tracks during which ticks the agent should leave Anthrax traces on the AnthraxLayer.
    /// </summary>
    private readonly List<long> _leaveAnthraxTraceTicks = new();
    
    /// <summary>
    ///     The total number of infections of the agent throughout the simulation.
    /// </summary>
    public int InfectedTotalCounter { get; set; }
    
    /// <summary>
    ///     A unique identifier of the agent.
    /// </summary>
    public Guid ID { get; set; }
    
    /// <summary>
    ///     The current position of the agent.
    /// </summary>
    public Position Position { get; set; }
    
    /// <summary>
    ///     A collection of positions that make up the agent's movement trajectory.
    /// </summary>
    private readonly List<Position> _positions = new();
    
    #endregion

    #region Initialization

    /// <summary>
    ///     
    /// </summary>
    /// <param name="layer">The layer type that registers this agent.</param>
    /// <exception cref="ArgumentException">Thrown if land type spawning probabilities do not add up to 1.0</exception>
    public void Init(AnimalLayer layer)
    {
        InfectedCounter = 0;
        InfectedTotalCounter = 0;
        Layer = layer;
        Energy = RandomHelper.Random.NextDouble(SpawnMinEnergy, SpawnMaxEnergy);
        State = AnimalState.RandomMove;

        if (SpawnWoodlandProbability + SpawnSavannaProbability != 1.0)
        {
            throw new ArgumentException("Spawning probabilities must add up to 1.0");
        }

        // no position set from kudu.csv -> choose random position according to land-type spawn probability
        if (Latitude == 0.0 || Longitude == 0.0)
        {
            Position = LandscapeLayer.GetRandomPositionForLandscapeType(RandomHelper.SmallerThan(SpawnWoodlandProbability) ?
                // spawn on Woodland
                new List<LandscapeType> {LandscapeType.Woodland} :
                // Spawn on Savanna   
                new List<LandscapeType> {LandscapeType.Savanna});
        }
        else
        {
            // position defined in csv file
            Position = Position.CreateGeoPosition(Longitude, Latitude);
        }
    }

    #endregion

    #region Tick

    public void Tick()
    {
        StoreTickResult = false;
        if (Layer.Context.CurrentTick == Layer.Context.MaxTicks || Layer.Context.CurrentTick == 1)
        {
            // in the first/last sim tick store the agent data regardless of anthrax infections so, we have 
            // the full set of agents in the output.
            StoreTickResult = true;
        }

        switch (State)
        {
            // Movement
            case AnimalState.RandomMove:
            {
                // in case we are on the wrong land type! Quickly move to the nearest comfortable area.
                // this can happen after visiting an water source, or after initialization.
                if (!_preferredLandTypes.Contains(LandscapeLayer.GetTypeForPosition(Position)))
                {
                    var nearestPreferredFeature = LandscapeLayer.FindNearestLandAreaOfType(Position, _preferredLandTypes.First());  // TODO replace First() call with some land type decision logic
                    var posInPreferredArea =nearestPreferredFeature.VectorStructured.Geometry.RandomPositionFromGeometry();
                    var bearing = Position.GetBearing(posInPreferredArea);
                    
                    // also update the distance, so we at most walk directly onto of the preferred landscape
                    // but never farther, which might lead us out of the perimeter!
                    var distance = Math.Min(GetMovementDistance(), Position.DistanceInMTo(posInPreferredArea));
                    Position = Layer.KuduEnvironment.MoveTowards(this, bearing, distance);
                }
                else
                {
                    // we are in a comfortable spot, let's move random, and check land type/distance to water
                    var moved = false;
                    do
                    {
                        var bearing = RandomHelper.Random.NextDouble(0, 360);
                        var distance = GetMovementDistance();
                        var target = Position.CalculateRelativePosition(bearing, distance);
                    
                        // is target still in area of KNP?
                        if (!Perimeter.IsPointInside(target))
                        {
                            continue;
                        }
                    
                        // is target of same category as our current position?
                        var targetType = LandscapeLayer.GetTypeForPosition(target);
                        if (targetType == LandscapeType.Woodland)
                        {
                            // every thing is fine.
                        }
                        else if (targetType == LandscapeType.Savanna)
                        {
                            if (RandomHelper.SmallerThan(MovementOnSavanna))
                            {
                                // it's our low probability we allow the animal on the savanna area
                            }
                            else
                            {
                                continue;
                            }
                        }
                        else
                        {
                            // unknown land type, try again!
                            continue;
                        }

                        // if the agent is to far from a water source walk towards it!
                        var nearestWaterSource = WaterLayer.GetNearestWaterSource(Position);
                        var g = new Point(Position.X, Position.Y);
                        //var d = nearestWaterSource.VectorStructured.Geometry.Distance(g);
                        var ps = DistanceOp.NearestPoints(g, nearestWaterSource.VectorStructured.Geometry);
                        var wPos = new Position(ps[1].X, ps[1].Y);
                        var d = GeoPositionExtension.DistanceInMTo(Position, wPos);
                    
                        if (d > MaxDistanceFromWaterInM)
                        {
                            bearing = Position.GetBearing(wPos);
                            distance = Math.Min(d, distance);
                        }
                    
                        Position = Layer.KuduEnvironment.MoveTowards(this, bearing, distance);
                        moved = true;
                    } while (!moved);
                }
            
                if (Energy < 15)
                {
                    State = AnimalState.SearchForWater;
                }
                break;
            }
            case AnimalState.SearchForWater:
            {
                // Energy is low, so look for water
                var waterSources = WaterLayer.Explore(Position.PositionArray, 50000).ToList();
                if (waterSources.Any())
                {
                    // Explore() is not sorted by distance, so we need to sort them first.
                    IVectorFeature nearestWaterSource = new VectorFeature();
                    var nearestWaterDistance = Double.MaxValue;
                    foreach (var w in waterSources)
                    {
                        var myPositionAsPoint = new Point(Position.X, Position.Y);
                        var d = w.VectorStructured.Geometry.Distance(myPositionAsPoint);
                    
                        //var loc = (Point) w.VectorStructured.Geometry;
                        //var t = new Position(loc.X, loc.Y);
                        //var d = t.DistanceInMTo(Position);

                        if (d < nearestWaterDistance)
                        {
                            nearestWaterDistance = d;
                            nearestWaterSource = w;
                        }
                    }
                
                    // Get coordinates of the nearest water source...
                    //var nearestWaterSource = waterSources.First();
                    // @todo: RandomPositionFromGeometry() is needed since die direct way for LINESTRINGs would be the 
                    // requires the nearest POINT on the target geoemtry, but we don't have that from the API. So we just
                    // use a random position somewhere on the linestring, so at least we go in the right direction...
                    var waterSourceLocation = nearestWaterSource.VectorStructured.Geometry.RandomPositionFromGeometry();
                    var target = new Position(waterSourceLocation.X, waterSourceLocation.Y);
                
                    // Math.Min() of random move distance and distance to water source, so we can reach the sight!
                    var distance = Math.Min(target.DistanceInMTo(Position), GetMovementDistance());
                
                    // ... and change the agent's bearing such that it looks in the direction of the water source
                    var bearing = Position.GetBearing(target);
                    Position = Layer.KuduEnvironment.MoveTowards(this, bearing, distance);

                    // If the agent in close the the water source, increase its energy and change bearing
                    if (target.DistanceInMTo(Position) < 20)
                    {
                        State = AnimalState.Drinking;
                    }
                }
                else
                {
                    Console.WriteLine("No water in area");
                }

                break;
            }
            case AnimalState.Drinking:
            {
                Energy += 50;
                if (Energy > 101)
                {
                    State = AnimalState.RandomMove;
                }

                break;
            }
            default:
                Console.WriteLine("Unknown agent state in Kudu agent.");
                break;
        }
        
        // Energy drops each tick
        Energy -= 1;

        if (OutputAgentTrack)
        {
            _positions.Add(Position.Copy());
        }

        // Infection
        if (AnthraxLayer.IsInRaster(Position) && AnthraxLayer.GetValue(Position) > 0)
        {
            var beta = AnthraxLayer.GetValue(Position) * AnthraxInfectionProbability;

            if (RandomHelper.SmallerThan(beta))
            {
                StoreTickResult = true;
                InfectedCounter += 1;
                InfectedTotalCounter += 1;
                
                var deathOccursInTicks = RandomHelper.Random.NextInteger(MinInfectionDurationInTicks,
                    MaxInfectionDurationInTicks + 1);
                var x = Layer.Context.CurrentTick + deathOccursInTicks;

                _leaveAnthraxTraceTicks.Add(x);
            }
        }

        // Leave Anthrax trail on the Map / animal dies
        if (InfectedCounter > 0)
        {
            if (_leaveAnthraxTraceTicks.Contains(Layer.Context.CurrentTick))
            {
                StoreTickResult = true;
                InfectedCounter -= 1;
                _leaveAnthraxTraceTicks.Remove(Layer.Context.CurrentTick);

                if (AnthraxLayer.IsInRaster(Position))
                {
                    AnthraxLayer[Position] += 1;
                }
            }
        }

        // Leave Movement trace for heatmap
        if (KuduMovement.IsInRaster(Position))
        {
            KuduMovement[Position] += 0.1;
        }

        // On last tick export movement of this agent as GeoJSON LineString
        if (OutputAgentTrack && Layer.GetCurrentTick() == Layer.Context.MaxTicks)
        {
            var featureCollection = new FeatureCollection();
            var coors = new List<Coordinate>();

            foreach (var p in _positions)
            {
                coors.Add(new Coordinate(p.X, p.Y));
            }
            
            var lineString = new LineString(coors.ToArray());
            featureCollection.Add(new Feature(lineString, new AttributesTable()));
            var write = new GeoJsonWriter().Write(featureCollection);
            File.WriteAllText($"Kudu_path_{ID}.geojson", write);
        }
    }

    #endregion
    
    #region Methods

    /// <summary>
    ///     Generates a random movement distance for the current tick.
    /// </summary>
    /// <returns>The movement distance the agent will traverse during the current tick.</returns>
    private double GetMovementDistance()
    {
        return RandomHelper.Random.NextDouble(MinMovementPerTickInM, MaxMovementPerTickInM);
    }

    #endregion
}